coreclient/src/outbound_service/timed_tasks_queue.rs

```diff
@@ -0,0 +1,135 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use chrono::{DateTime, Utc};
+use sqlx::{
+    Database, Decode, Encode, Sqlite, Type, encode::IsNull, error::BoxDynError,
+    sqlite::SqliteTypeInfo,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum TaskKind {
+    KeyPackageUpload,
+}
+
+impl TaskKind {
+    pub(super) fn default_interval(&self) -> chrono::Duration {
+        match self {
+            TaskKind::KeyPackageUpload => chrono::Duration::weeks(1),
+        }
+    }
+}
+
+pub(crate) struct TimedTaskQueue {
+    due_at: DateTime<Utc>,
+    kind: TaskKind,
+}
+
+impl TimedTaskQueue {
+    pub(crate) fn new_key_package_upload_task(due_at: DateTime<Utc>) -> Self {
+        Self {
+            due_at,
+            kind: TaskKind::KeyPackageUpload,
+        }
+    }
+}
+
+mod persistence {
+    use sqlx::{SqliteExecutor, query, query_scalar};
+    use tracing::debug;
+    use uuid::Uuid;
+
+    use super::*;
+
+    impl TimedTaskQueue {
+        pub(crate) async fn enqueue(&self, executor: impl SqliteExecutor<'_>) -> sqlx::Result<()> {
+            debug!(
+                ?self.due_at, ?self.kind, "Enqueueing timed task"
+            );
+
+            query!(
+                "INSERT INTO timed_tasks_queue
+                    (due_at, task_kind)
```

Comment:
I think this is where it is important to have a primary key on `task_kind` to avoid scheduling the same task several times. After it is added, this call should be an upsert.